controllers/reconcilers/kserve_raw_route_reconciler.go

```diff
@@ -0,0 +1,195 @@
+	return &KserveRawRouteReconciler{
+		client:         client,
+		routeHandler:   resources.NewRouteHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveRawRouteReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling Generic Route for Kserve InferenceService")
+
+	// Create Desired resource
+	desiredResource, err := r.createDesiredResource(ctx, log, isvc)
+	if err != nil {
+		return err
+	}
+
+	// Get Existing resource
+	existingResource, err := r.getExistingResource(ctx, log, isvc)
+	if err != nil {
+		return err
+	}
+
+	// Process Delta
+	if err = r.processDelta(ctx, log, desiredResource, existingResource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *KserveRawRouteReconciler) Delete(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Deleting Kserve inference service generic route")
+	return r.routeHandler.DeleteRoute(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace})
+}
+
+func (r *KserveRawRouteReconciler) Cleanup(_ context.Context, _ logr.Logger, _ string) error {
+	// NOOP - resources are deleted together with ISVCs
+	return nil
+}
+
+func (r *KserveRawRouteReconciler) createDesiredResource(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) (*v1.Route, error) {
+	var err error
+	enableAuth := false
+	if enableAuth, err = strconv.ParseBool(isvc.Labels[constants.LabelEnableAuthODH]); err != nil {
+		enableAuth = false
+	}
+	createRoute := false
+	if val, ok := isvc.Labels[constants.KserveNetworkVisibility]; ok && val == constants.LabelEnableRoute {
+		createRoute = true
+	}
+	if !createRoute {
+		log.Info("InferenceService does not have label '" + constants.KserveNetworkVisibility + "' annotation" +
+			" set to '" + constants.LabelEnableRoute + "'. Skipping route creation")
+		return nil, nil
+	}
+
+	// Fetch the service with the label "serving.kserve.io/inferenceservice=isvc.Name" in the isvc namespace
+	serviceList := &corev1.ServiceList{}
+	labelSelector := client.MatchingLabels{"serving.kserve.io/inferenceservice": isvc.Name}
+	err = r.client.List(ctx, serviceList, client.InNamespace(isvc.Namespace), labelSelector)
+	if err != nil || len(serviceList.Items) == 0 {
+		log.Error(err, "Failed to fetch service for InferenceService", "InferenceService", isvc.Name)
+		return nil, err
+	}
+	targetServiceName := serviceList.Items[0].Name
+
+	desiredRoute := &v1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      isvc.Name,
+			Namespace: isvc.Namespace,
+			Labels: map[string]string{
+				"inferenceservice-name": isvc.Name,
+			},
+		},
+		Spec: v1.RouteSpec{
+			To: v1.RouteTargetReference{
+				Kind:   "Service",
+				Name:   targetServiceName,
+				Weight: pointer.Int32(100),
+			},
+			Port: &v1.RoutePort{
+				TargetPort: intstr.FromInt(isvcServicePort),
```

Comment:
I don't think we can hard-code to port `8888`.

Caikit runtimes are using `8080`, OVMS is using `8888`, tgis is using `8033`, and vLLM is using `8080`. If we add new runtimes, other ports may be coming. So, it is better to resolve the port.

...But the Route is pointing to the `Service`. So, I wonder if it shouldn't be the common http port?

Reply:
https://github.com/opendatahub-io/kserve/pull/419/files#diff-401b6cf36f71bdda9ab1593723f98632b0e7ee8d3a98151d65439e13d8d2ea9fR264 
the oauth exposed port is 8443, the oauth proxy will in turn route to the correct runtime port

Reply:
You are right: when auth is enabled there is some code below that will change this to `8443`. However, this line seems to be covering the case when auth is disabled. For that case, it looks like the Route will be pointing to an erroneous port.